test/vesting.js

```diff
@@ -0,0 +1,347 @@
+    let durationSeconds2 = 200;
+    let amount2 = 2_000_000;
+
+    it("Should correctly vest", async function () {
+      await ethers.provider.send("evm_setAutomine", [true]);
+      let VestingFactory = await ethers.getContractFactory(
+        "ArrowVestingFactory"
+      );
+      let Token = await ethers.getContractFactory("ArrowToken");
+      let name = "Arrow Token";
+      let symbol = "ARROW";
+      arrowToken = await Token.deploy(
+        ethers.BigNumber.from(100_000_000),
+        name,
+        symbol
+      );
+      await arrowToken.deployed();
+      let VestingBase = await ethers.getContractFactory("ArrowVestingBase");
+
+      [_, beneficiary1, beneficiary2] = await ethers.getSigners();
+      vestingFactory = await VestingFactory.deploy();
+      await vestingFactory.deployed();
+
+      // get block timestamp
+      const blockNum = await ethers.provider.getBlockNumber();
+      const block = await ethers.provider.getBlock(blockNum);
+      startTimestamp = block.timestamp + 200;
+      startTimestamp2 = block.timestamp + 250;
+
+      // use callStatic to get what the address would be when the transaction is created
+      var vestingWalletAddress =
+        await vestingFactory.callStatic.createVestingSchedule(
+          beneficiary1.address,
+          ethers.BigNumber.from(startTimestamp),
+          ethers.BigNumber.from(durationSeconds)
+        );
+      // actually create the transaction so that the child contract can be deployed on chain
+      var txn = await vestingFactory.createVestingSchedule(
+        beneficiary1.address,
+        ethers.BigNumber.from(startTimestamp),
+        ethers.BigNumber.from(durationSeconds)
+      );
+      await txn.wait();
+      vestingContract = await VestingBase.attach(vestingWalletAddress);
+
+      // use callStatic to get what the address would be when the transaction is created
+      vestingWalletAddress =
+        await vestingFactory.callStatic.createVestingSchedule(
+          beneficiary2.address,
+          ethers.BigNumber.from(startTimestamp2),
+          ethers.BigNumber.from(durationSeconds2)
+        );
+      // actually create the transaction so that the child contract can be deployed on chain
+      txn = await vestingFactory.createVestingSchedule(
+        beneficiary2.address,
+        ethers.BigNumber.from(startTimestamp2),
+        ethers.BigNumber.from(durationSeconds2)
+      );
+      await txn.wait();
+      vestingContract2 = await VestingBase.attach(vestingWalletAddress);
+
+      /*---------------------------------Test Begin---------------------------------**/
+      expect(await arrowToken.balanceOf(beneficiary1.address)).to.equal(0);
+      expect(await arrowToken.balanceOf(beneficiary2.address)).to.equal(0);
+
+      // Load contracts with tokens
+      await arrowToken.transfer(vestingContract.address, amount);
+      await arrowToken.transfer(vestingContract2.address, amount2);
+
+      console.log("√ Pass - Load contracts with tokens");
```

Comment:
For my benefit, what's the purpose of these log functions? Don't the `expect` calls do the job of telling you whether a part of the test has passed/failed?

Reply:
`expect` only checks for the constraints passed in, it does not log anything on the console unless `expect` fails. 

The passing logs you see are generated by `it` and `describe` modules. For this particular test, though, I had to put everything under one `it` module, since vesting is a continuous process with one test contingent upon the execution of a previous test (i.e. the vesting balance, the block number, etc).

Without these custom logs, you'd only see one line on the console to indicate whether the whole thing has passed or failed, which would makes debugging harder. In other words, these custom logs were made to mimic the testing behavior as if we have broken down the whole thing into more discrete testing modules.

Reply:
I don't think the comments add anything, because they're just mimicking the behaviour of the `expect` above each one. 

If the test is passing, I don't really care about the console output, and if it's failing, I'm going to go right the the `expect` that failed and start debugging from there.

I'd consider replacing them with an `expect(condition, message)` custom error message, which would be more helpful when things are failing and more localised than comments.